java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -259,6 +260,22 @@ public String toString() {
     return "{" + values.toString() + "}  {" + constraints.toString() + "}" + " { " + stack.toString() + " }";
   }
 
+  public ProgramState addConstraintTransitively(SymbolicValue symbolicValue, Constraint constraint) {
+    final ProgramState[] ps = {addConstraint(symbolicValue, constraint)};
+    knownRelations()
+      .filter(rsv -> rsv.isEquality() && (rsv.getLeftOp() == symbolicValue || rsv.getRightOp() == symbolicValue))
+      .map(rsv -> rsv.getLeftOp() == symbolicValue ? rsv.getRightOp() : rsv.getLeftOp())
+      .forEach(sv -> ps[0] = ps[0].addConstraint(sv, constraint));
```

Comment:
Unfortunately, stream API doesn't provide fold-left operation, and reduce method would require combiner to allow merging of two intermediate ProgramStates. But I understand that abusing arrays for this is bit of a hack, so I rewrote it by having values collected.